docs/controller.md

```diff
@@ -0,0 +1,108 @@
+# Architecture Document for Controller Component
+
+## Description
+The Controller is the component giving meaning to the [pipeline](#structure) file given by the Monitor. It receives [pipelines](#pipeline), parse them into a set of [actions](#actions) and send these actions sequentially to the Scheduler, for each of these actions, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error, it can be done just by sending the logs from the Agent for instance. Thanks to these information, the Scheduler is able to provide information about a pipeline state to anyone (the Monitor or any other client).
```

Comment:
```suggestion
The Controller is the component giving meaning to the [pipeline](#structure) file given by the Monitor. 

It receives [pipelines](#pipeline), parses them into a set of [actions](#actions) and send these actions sequentially to the Scheduler. 

For each submitted action, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error. 

This can be done by sending the logs from the Agent for instance. Thanks to these information, the Scheduler is able to provide information about a pipeline state to anyone (the Monitor or any other client).
```

Now a few comments:

- What do you mean by `giving meaning`?
- The scheduler will only notify the controller about an action status. The controller is responsible for assembling that data back into an over pipeline status (e.g. a pipeline failed when one of its actions failed).
- How the agent notifies the scheduler about an action status/completion is out of the controller scope. As a matter of fact, it should not know about it.
- I don't think that a scheduler should even know about the concept of a pipeline. That's one of the Controller roles: Split a pipeline into atomic units (actions) to submit to the scheduler.

Reply:
In fact, the following sentence:
> For each submitted action, the Scheduler **must** notify the Controller when a job has been scheduled and has been completed successfully or encountered an error. 

does not correspond to the correct behaviour of the controller and the scheduler. The scheduler just returns the agent's output, and then it's up to the controller to use conventions to determine whether or not the job has failed, and then orchestrate the rest of the action (pipeline). 

Translated with DeepL.com (free version)

Reply:
I disagree with that. Only the agent knows if an action has failed or not, by virtue of getting the corresponding process exit status back. If the agent only returns logs back to the scheduler, you lose that fundamental piece of information. The controller not only may have no knowledge about the semantics of an action log, but letting it interpret and parse potentially very large logs is not a scalable option.
Moreover, actions may produce no logs at all.

Reply:
Done, we modified the description and added a dedicated status code in the protobuf for the controller/scheduler interface